### ISSUE TITLE
Provide .editorconfig support

### ISSUE THREAD

**Issue by angelozerr** [eclipse-platform/eclipse.platform.ui#877]
Today `.editorconfig` is supported directly in vscode, IJ, etc

It should be really nice that Eclipse IDE provides this support too. It exists a discussion at https://bugs.eclipse.org/bugs/show_bug.cgi?id=457046 

And I implemented a long time ago https://github.com/angelozerr/ec4e which works with genereic editor but cannot apply all settings.

**Comment by laeubi** [comment]
@angelozerr do you think you can prepare a PR for this to contribute the code to eclipse? I'll then open a CQ for this,

**Comment by mickaelistria** [comment]
The edition part seems to require TM4E. Do we want Platform to start depending on TM4E ? I personally wouldn't mind, but I think it should be discussed a bit more broadly.
If we don't want it, then things like langauge configuration and highlighting would have to be removed for the moment.

**Comment by jukzi** [comment]
some questions: 
a) What happens when the .editorconfig contradicts the workspace settings?
b) What happens when the .editorconfig in the project folder contradicts the project settings?
c) When the projects setting are edited via UI will the .editorconfig reflect the change?
d) does "searching" for the .editorconfig use file system or eclipse workspace api?

**Comment by angelozerr** [comment]
At first I would like to say that ec4e is based on https://github.com/ec4j/ec4j an editorconfig parser that I have implemented which can provide the required information that we need for IDE (AST offsets of section, fault tolerant parser,etc) compared to https://github.com/editorconfig/editorconfig-core-java which don't give the capability to give some offsets.

The official site https://editorconfig.org/ references today the ec4j.

The ec4e should work well with Generic Editor but not correctly with old editor. Here a demo where you can see wizard in action, codemining to show number of files which matche the section, syntax validation of .editorconfig and the most important apply of setting for matched files:

![EditorConfigDemo](https://user-images.githubusercontent.com/1932211/190447190-a8183175-77bc-414a-ab98-c78c60ddebd6.gif)

> The edition part seems to require TM4E. Do we want Platform to start depending on TM4E ? I personally wouldn't mind, but I think it should be discussed a bit more broadly.

Indeed, perhaps TM4E should take care of that?

> a) What happens when the .editorconfig contradicts the workspace settings?

It is the .editorconfig which should have the priority. But any improvement are possible.

> b) What happens when the .editorconfig in the project folder contradicts the project settings?

Again, it is the .editorconfig which should have the priority. But any improvement are possible.

> c) When the projects setting are edited via UI will the .editorconfig reflect the change?

No.

> d) does "searching" for the .editorconfig use file system or eclipse workspace api?

ec4j provide some abstract resource class to retrieve .editorconfig. ec4e implement them with Eclipse API IResource https://github.com/angelozerr/ec4e/tree/master/org.eclipse.ec4e/src/main/java/org/eclipse/ec4e/internal/resource

**Comment by merks** [comment]
I'm not sure it's a good idea to have more circular dependencies in the platform via TM4E.   Ssuch an addition would force Java 17 BREE into the platform, which is yet another issue to consider carefully.

**Comment by laeubi** [comment]
> Ssuch an addition would force Java 17 BREE into the platform, which is yet another issue to consider carefully.

Build is already Java 17, EPP is already Java 17, so having one bundle (we don't need to put the code into an exiting plugin) would not change much?

> I'm not sure it's a good idea to have more circular dependencies in the platform via TM4E

@mickaelistria  already mentioned one might be able to not use TM4E beside that, if we just have an `org.eclipse.text.editorconfig` bundle it should not be much of an issue...

**Comment by akurtakov** [comment]
Having circular dependency between tm4e and platform.releng build is a total no-go. Java 17 is not an issue for me (it's probably time to admit that Java 11 support other than producing Java 11 bytecode is of zero interest for my company team now).

**Comment by laeubi** [comment]
> Having circular dependency between tm4e and platform.releng build is a total no-go

Maybe we have a different understanding of a "circular dependency" but for me a circular dependency is where `A > B > A` what can not be resolved, while here we have `A > C < B` what can be resolved... 

Anyways @angelozerr can you explain where TM4E is used and if it can be either optional or replaced with other techniques?

**Comment by akurtakov** [comment]
> > Having circular dependency between tm4e and platform.releng build is a total no-go
> 
> Maybe we have a different understanding of a "circular dependency" but for me a circular dependency is where `A > B > A` what can not be resolved, while here we have `A > C < B` what can be resolved...

It's a releng "circular dependency" between p2 repos -  platform  -> tm4e -> platform

> 
> Anyways @angelozerr can you explain where TM4E is used and if it can be either optional or replaced with other techniques?

**Comment by vogella** [comment]
We have some circular dependencies already IIRC. The EMF edit plug-in IIRC.

**Comment by laeubi** [comment]
> We have some circular dependencies already IIRC. The EMF edit plug-in IIRC.

The whole platform build is a "circular dependency" ;-)

Beside that, no one forces us to use the *latest snapshot repository* of TM4E ... we can simply use the *latest release* of TM4E, of course then might only use features of TM4E - 1 but this should not be an issue here...

**Comment by akurtakov** [comment]
Please, the fact that we have many issues doesn't mean we should add more! Let's agree to not make platform build worse than its current terrible state.

**Comment by akurtakov** [comment]
So to move the discussion in the right direction - TM4E is used for the editor of .editorconfig files which means that applying/supporting existing .editorconfig files can go in without the requirement on TM4E and I propose the discussion to go towards that goal. Having smart editor for .editorconfig files can stay separate.

**Comment by laeubi** [comment]
Even though I don't know how it works, but Eclipse sometimes recommend me to install things from the marketplace for "unknown" extensions.

So if we can integrate the `.editorconfig` itself, and link it to a market-place entry then it could be installed afterwards by a seperate project that support the editing of files.... @angelozerr would that be feasible for you?

**Comment by akurtakov** [comment]
Marketplace entry that has a tag fileextension_editorconfig should be everything needed for MPC to suggest installing the editor.

**Comment by angelozerr** [comment]
Dont worry I can remove TM4E which provides syntax coloration and some features with langage configuration wich provides machine bracket for .editorconfig file but the other factures like validation codemining complerion for .editorconfig is not linked to TM4E.

The only required dépendance is ec4j which is on maven central.

Keep in mind the apply of .editorconfig works correctly with only generic editor and not with some editor like jdt java editor.

More I need to use java reflection to add a custom IPreferenceStore to generic editor which is not very clean.

To be clean generic editor should provide perhaps an extension point to add an ipreferencestore

**Comment by angelozerr** [comment]
For .editorconfig syntax coloration I think TM4E should provide the textmate for .editorconfig file.

**Comment by angelozerr** [comment]
I have just one question for ec4j dependency, can I create a PR with plusieurs which have à lib folder which contains ec4j jar?

We could manage that properly in an another PR?  What do you think about that?

**Comment by akurtakov** [comment]
It's fine to have first draft with ec4j being internal dependency. Before accepting we should have it added to the [target platform](https://github.com/eclipse-platform/eclipse.platform.releng.aggregator/blob/master/eclipse.platform.releng.prereqs.sdk/eclipse-sdk-prereqs.target) referring to the [maven artifact](https://mvnrepository.com/artifact/org.ec4j.core/ec4j-core) directly.

**Comment by laeubi** [comment]
> I have just one question for ec4j dependency, can I create a PR with plusieurs which have à lib folder which contains ec4j jar?

You can add maven dependencies directly to the target and use is like any regular bundle, so no need for workaround like lib folders!

**Comment by angelozerr** [comment]
I have created a draft PR at https://github.com/eclipse-platform/eclipse.platform.text/pull/89 without TM4E but we need to do some PR before, see https://github.com/eclipse-platform/eclipse.platform.text/pull/89#issuecomment-1250040908

**Comment by p-bakker** [comment]
Imho having .editorconfig support consists of several things.

The first level is parsing the file and applying the correct settings from it to the editor.

The next step is acting on those preferences: while I think the indent_style and indent_size do map on existing, workspace-level settings (that are currently already taken into account in most of not all editors), settings like `trim_trailing_whitespace` and `insert_final_newline` do not (to my knowledge). These settings exist for the JDT Java editor, but aren't generic, workspace-wide settings available to/implemented in all editors.

I for one would expect that if Eclipse would support .editorconfig, it would also mean that all the settings one can set via .editorconfig would also be supported in all editors (although personally I'd be ok with just support in the generic editor)

Third level would be rich editing support for .editorconfig files

**Comment by p-bakker** [comment]
> some questions:
> a) What happens when the .editorconfig contradicts the workspace settings?
> b) What happens when the .editorconfig in the project folder contradicts the project settings?
> c) When the projects setting are edited via UI will the .editorconfig reflect the change?

Per definition, .editorconfig overrides workspace- and project-level settings: it defines settings on a per-fileType basis, looking for the closets applicable settings by looking for the closets .editorconfig file (containing relevant settings) from the current filled location up through its parent directory structure, until either matching settings have been found or a .editorconfig file is encountered that has `root = true`

**Comment by laeubi** [comment]
> for one would expect that if Eclipse would support .editorconfig, it would also mean that all the settings one can set via .editorconfig would also be supported in all editors

Given that Eclipse is extensible "all" might never be achievable ... so starting with "one" that is recommended to be used for new editors seems suitable to me.

**Comment by angelozerr** [comment]
> I for one would expect that if Eclipse would support .editorconfig, it would also mean that all the settings one can set via .editorconfig would also be supported in all editors (although personally I'd be ok with just support in the generic editor)

> Given that Eclipse is extensible "all" might never be achievable ... so starting with "one" that is recommended to be used for new editors seems suitable to me.

It is an important question, should .editorconfig works with any editor or with just genereic editor? The answer will determine where the extension point which create and add an IPreferenceStore into the editor (in texteditor extension point or in genereric extension point).

**Comment by laeubi** [comment]
Just assume I have an editor that supports Grahical editing of a binary file format, has "trim_trailing_whitespace" any meaning?

**Comment by laeubi** [comment]
By the way, as far as I know no one prevents others to also looking at the generic editor extension point later on...

**Comment by jukzi** [comment]
> Again, it is the .editorconfig which should have the priority. But any improvement are possible.

whatever takes precedence should be shown in the properties file of any resource:
1. the effective used file 
2. the effective setting
![image](https://user-images.githubusercontent.com/51790620/191028595-e6ff749c-19cd-4aae-97f1-bedbaca74c53.png)

**Comment by laeubi** [comment]
> whatever takes precedence should be shown in the properties file of any resource:

I don't think that this is mandatory, any editor can ignore these defaults these are just resources settings. Still it would be useful if there would be an extension point to inform the user about that these settings are overridden somewhere else.

**Comment by angelozerr** [comment]
I agree with your all comments, but to be honnest with you I could not do that in the same PR and I'm not sure that I will have time to implement all of your ideas. My main goal is to provide .editorconfig basicly and after we could improve it step by step.

**Comment by angelozerr** [comment]
To provide a clean API to contribute to generic editor with custom preferences store (with extension point), I need thoses 2 PRs:

 * https://github.com/eclipse-platform/eclipse.platform.text/pull/107
 * https://github.com/eclipse-platform/eclipse.platform.text/pull/108

**Comment by mickaelistria** [comment]
I don't think Eclipse Platform will ever provide ECJ support directly, but instead would expect it to be provided by another plugin from another project. So I'm closing the issue here.
But the goal is still interesting and proposals to improve ability to add support for external formatting options from 3rd party plugins are still very welcome.

**Comment by laeubi** [comment]
@angelozerr are you still interested in bringing editorconfig support to eclipse?

I just got hit by the problem that an project uses editorconfig and it interferes with my usual settings so I need to manually reformat the code (and be careful Eclipse don't break it). And as many things have changed over time (e.g. Java 17 is no issue anymore at all) I think it might make sense to pick up this again.

I also reviewed a bit the code you shared here, the PR and the github repository and I think we have different topics that can be addressed independently:

1. Make Eclipse capable of **reading** `.editorconfig` files, this could be done for example https://github.com/angelozerr/ec4j can you give a status here what is the current state, e.g. is it still maintained, releases to maven central for example and so on? Any special dependencies we should be aware of? Or do you propose any alternative parser as of today?
2. Add an editor for `.editorconfig` files, I think this is somehow part of https://github.com/angelozerr/ec4e and could stay there independently but we might want to include it into EPPs (FYI @jonahgraham )
3. Take editor config into account for generic editor, that's also part of https://github.com/angelozerr/ec4e and would be good to be ported/included here by default (no additional UI required as one can edit the file directly)
4. Take editorconfig into account for JDT: Here my idea would be to either have a gobal preference "Take editorconfig into account" on the formater settings, or have a setting in the formater profile "allow override by editorconfig" so it reads some settings from there and overrides the ones specified in the profile, most notable the used whitespace character could be a very first and basic step
5. Take encoding settings into account (as mentioned by @jukzi previously) this is independent from editors and/or formatter and one seem to plug into the filetypes here that already can provide an encoding for different filetypes (but in a static way).

**Comment by laeubi** [comment]
I now added a very basic proof-of-concept that already works very well in my setup that simply overrides the tab char:

- https://github.com/eclipse-jdt/eclipse.jdt.ui/pull/1872

**Comment by angelozerr** [comment]
> @angelozerr are you still interested in bringing editorconfig support to eclipse?

Even if the topic is super interesting, unfortunately I will have no time to do that again -(

> Make Eclipse capable of reading .editorconfig files, this could be done for example https://github.com/angelozerr/ec4j can you give a status here what is the current state, e.g. is it still maintained, releases to maven central for example and so on? Any special dependencies we should be aware of? Or do you propose any alternative parser as of today?

https://github.com/angelozerr/ec4j is my repository I used when I have created this project. I have created this project because the  only official project which .editorconfig parser in Java was https://github.com/editorconfig/editorconfig-core-java#readme was not
helpfull to integrate it in an IDE, because it doesn't take care of AST node.

It is the one reason why I have created this project and today the official project is https://github.com/ec4j/ec4j The very good news is that this parser is today mature and robust:

 * you can see at https://editorconfig.org/ that ec4j is referenced
 * JetBrains has switchted from https://github.com/editorconfig/editorconfig-core-java to ec4j (see https://github.com/search?q=repo%3AJetBrains%2Fintellij-community%20ec4j&type=code)
 * ec4j is used in several tooling https://github.com/ec4j and maintained by @ppalaga 

> Add an editor for .editorconfig files, I think this is somehow part of https://github.com/angelozerr/ec4e and could stay there independently but we might want to include it into EPPs (FYI @jonahgraham )

This project is based on https://github.com/angelozerr/ec4j and uses

 * tm4e for supporting syntax coloration -> platform UI cannot add this dependency to tm4e, I think tm4e should include the textmate grammar from .editorconfig
 * generic editor to add codelens, validation, hover, completion -> those all code could be hosted in the platform 

But the most important thing is apply the editorconfig properties to a given editor (for standard editor and generic editor both). It is the hardest topic. I spend a lot of time to find a generic solution (used by generic editor orother component), And I think the first topic that you will need to work is https://github.com/eclipse-platform/eclipse.platform.text/pull/130

> Take editor config into account for generic editor, that's also part of https://github.com/angelozerr/ec4e and would be good to be ported/included here by default (no additional UI required as one can edit the file directly)

It is the easy part and you could copy/paste ec4e inside Platform UI (without the textmate grammar)

> Take editorconfig into account for JDT: Here my idea would be to either have a gobal preference "Take editorconfig into account" on the formater settings, or have a setting in the formater profile "allow override by editorconfig" so it reads some settings from there and overrides the ones specified in the profile, most notable the used whitespace character could be a very first and basic step

The first job to do is to provide an extension point used by the AbstractTextEditor https://github.com/eclipse-platform/eclipse.platform.text/pull/130

It is long and hard work, thanks so much @laeubi to work on this great features again!

**Comment by angelozerr** [comment]
@laeubi I think the full PR that you should see is https://github.com/eclipse-platform/eclipse.platform.text/pull/109/files but it is working only with generic editor if I remember

**Comment by angelozerr** [comment]
@laeubi about the editorconfig editor, you can see all features that I had implemented in https://github.com/angelozerr/ec4e/tree/master/org.eclipse.ec4e/src/main/java/org/eclipse/ec4e/internal (
codemining, completion, folding, hover , outline, validation and wizards)

Those feature uses the ec4j parser which build an AST nodes. codemining support is very nice because it displays codeming before section and you can click on it to perform the Search UI to retrieve all files which matches the section.

**Comment by angelozerr** [comment]
ec4j is available on maven central https://mvnrepository.com/artifact/org.ec4j you will need the ec4j core https://github.com/ec4j/ec4j

The ec4j parser isvery robust because it uses the official test of editorconfig https://github.com/ec4j/ec4j/tree/master/core
and we have a lot of test for the parser https://github.com/ec4j/ec4j/tree/master/core/src/test/java/org/ec4j/core

**Comment by BuZZ-dEE** [comment]
I would be great to have it, because "[EditorConfig](https://editorconfig.org/) is somewhat of an [industry-standard](https://github.com/zed-industries/zed/issues/8534#issue-2159040258) these days."

**Comment by angelozerr** [comment]
I agree but I have no time to work again on my old PR.

I think ec4j should be used as editorconfig parser since JetBrains use it.
